File: 2021/procon_java/src/Atcoder/abc_past/D_012.java
package Atcoder.abc_past;

import java.util.Scanner;

public class D_012 {

	public static final int INF = 999999;

	public static void main(String args[]) {

		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int M = sc.nextInt();

		int[][] d = new int[N][N];
		for(int i=0;i<N;i++) {
			for(int j=0;j<N;j++) {
				if(i!=j) {
					d[i][j]=INF;
				}else {
					d[i][j]=0;
				}
			}
		}

		for(int i=0;i<M;i++) {
			int a=sc.nextInt();
			int b=sc.nextInt();
			int t=sc.nextInt();
			d[a-1][b-1]=t;
			d[b-1][a-1]=t;
		}

		for(int a=0;a<N;a++) {
			for(int b=0;b<N;b++) {
				for(int c=0;c<N;c++) {
					d[b][c]=Math.min(d[b][c], d[b][a]+d[a][c]);
				}
			}
		}

		int ans=INF;
		for(int i=0;i<N;i++) {
			int ans_i=0;
			for(int j=0;j<N;j++) {
				ans_i=Math.max(ans_i, d[i][j]);
			}
			ans=Math.min(ans, ans_i);
		}

		System.out.println(ans);
	}

}

File: 2021/procon_python/src/atcoder/abc/past/D_012.py
INF=float('inf')
n,m=map(int,input().split())
d=[[INF for _ in range(n+1)] for _ in range(n+1)]

for i in range(1,n+1):
    d[i][i]=0

for i in range(m):
    a,b,t=map(int,input().split())
    d[a][b]=t
    d[b][a]=t


for a in range(n+1):
    for b in range(n+1):
        for c in range(n+1):
            d[b][c] = min(d[b][c],d[b][a]+d[a][c])

ans=INF
for i in range(1,n+1):
    ans=min(ans,max(d[i][1:]))
print(ans)